### PR TITLE
Hold read lock during startup shrink

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
validator startup time is painful - especially as # accounts grows. There will soon be an issue filed for this.
#### Summary of Changes
From verify_snapshot_bank, shrink slots reads index while holding the read lock.
Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
```
verify_snapshot_bank clean_us=57710861i shrink_all_slots_us=1097525i verify_bank_hash_us=19499188i verify_hash_us=2205i (master)
verify_snapshot_bank clean_us=57490783i shrink_all_slots_us= 622640i verify_bank_hash_us=19254628i verify_hash_us=2209i (#17308)
verify_snapshot_bank clean_us=58307796i shrink_all_slots_us= 358199i verify_bank_hash_us=19419787i verify_hash_us=2247i (this pr)
```

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -1947,31 +1947,43 @@ impl AccountsDb {
...
-                            // rewriting the storage entries.
-                            locked_entry.unref()
-                        } else {
-                            alive_total += stored_account.account_size as u64;
-                        }
-                        is_alive
+        let mut read_lock = None;
+        if is_startup {
+            read_lock = Some(self.accounts_index.get_account_maps_read_lock());
+        }
```
Comment (carllin) [inline runtime/src/accounts_db.rs:1953]: nit, a little more Rusty way of initialization: 

```
        let read_lock = if is_startup {
            Some(self.accounts_index.get_account_maps_read_lock())
        } else {
            None
        };
```

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -1947,31 +1947,43 @@ impl AccountsDb {
...
-                            // This pubkey was found in the storage, but no longer exists in the index.
-                            // It would have had a ref to the storage from the initial store, but it will
-                            // not exist in the re-written slot. Unref it to keep the index consistent with
-                            // rewriting the storage entries.
-                            locked_entry.unref()
-                        } else {
-                            alive_total += stored_account.account_size as u64;
-                        }
-                        is_alive
+        let mut read_lock = None;
```
Comment (carllin) [inline runtime/src/accounts_db.rs:1950]: nit: rename `read_lock` -> `accounts_index_map_lock`

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -1947,31 +1947,43 @@ impl AccountsDb {
...
+        if is_startup {
+            read_lock = Some(self.accounts_index.get_account_maps_read_lock());
+        }
+        let read_lock_ref = read_lock.as_ref();
```
Comment (carllin) [inline runtime/src/accounts_db.rs:1954]: nit: rename `read_lock_ref -> accounts_index_map_lock_ref`

File: runtime/src/accounts_db.rs [hunk 4]
```diff
@@ -1947,31 +1947,43 @@ impl AccountsDb {
...
-                            // It would have had a ref to the storage from the initial store, but it will
-                            // not exist in the re-written slot. Unref it to keep the index consistent with
-                            // rewriting the storage entries.
-                            locked_entry.unref()
-                        } else {
-                            alive_total += stored_account.account_size as u64;
-                        }
-                        is_alive
+        let mut read_lock = None;
+        if is_startup {
```
Comment (carllin) [inline runtime/src/accounts_db.rs:1951]: Maybe a comment on why this is more acceptable for startup, i.e., no replay/RPC thread to contend with?

**Review by carllin — APPROVED** [review]

**Pull request merged**